File: generated_code/GeminiPro/Python/002_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Splay Tree class implementing a dictionary-like set for integers.

    This implementation includes methods for insertion, deletion, and searching.
    The key feature of a splay tree is that recently accessed elements are moved
    to the root of thetree to optimize for temporal locality of reference.
    """

    class Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on the given node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on the given node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on the given node x, moving it to the root.
        """
        if not x:
            return
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-Zig case
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-Zig case
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-Zag case
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-Zag case (x == p.left and p == g.right)
                self._right_rotate(p)
                self._left_rotate(g)

    def insert(self, key):
        """
        Inserts a key into the tree.
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.
        """
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay it and return
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Key is not in the tree, insert a new node
        new_node = self.Node(key)
        new_node.parent = parent

        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def search(self, key):
        """
        Searches for a key in the tree.
        Splays the found node or the last-visited node (if not found) to the root.
        Returns True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Key not found, splay the last visited node if the tree is not empty
        if last_node:
            self._splay(last_node)
        return False

    def delete(self, key):
        """
        Deletes a key from the tree.
        This is done by first searching for the key (which splays it to the root),
        then removing it and joining the two resulting subtrees.
        """
        # Search for the key. This splays the node if found, or its parent if not.
        if not self.search(key):
            # key was not found. search() already splayed the would-be parent.
            return

        # If search returned True, the node with the key is now the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, the right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Find the maximum node in the left subtree. This will be the new root.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this maximum node to the root of the left subtree.
            # This is done by temporarily treating the left subtree as the whole tree.
            self.root = left_subtree
            left_subtree.parent = None
            self._splay(max_node)
            
            # Now, self.root is the new root (the splayed max_node).
            # Re-attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/002_SplayTree.java
/**
 * A complete, self-contained implementation of an integer set using a Splay Tree.
 * <p>
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, look-up and removal in
 * O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        public Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splay Operations ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The splay operation. Moves the node with the given key (or the last accessed
     * node on the search path) to the root of the tree.
     *
     * @param root The root of the tree (or subtree) to splay.
     * @param key  The key to search for and splay.
     * @return The new root of the tree after splaying.
     */
    private Node splay(Node root, int key) {
        // Base cases: root is null or key is present at root
        if (root == null || root.key == key) {
            return root;
        }

        // Key lies in the left subtree
        if (key < root.key) {
            // Key is not in tree, we are done
            if (root.left == null) {
                return root;
            }

            // Zig-Zig (Left-Left)
            if (key < root.left.key) {
                // Recursively bring the key as root of left-left
                root.left.left = splay(root.left.left, key);
                // Do first rotation for root
                root = rightRotate(root);
            }
            // Zig-Zag (Left-Right)
            else if (key > root.left.key) {
                // Recursively bring the key as root of left-right
                root.left.right = splay(root.left.right, key);
                // Do first rotation for root.left
                if (root.left.right != null) {
                    root.left = leftRotate(root.left);
                }
            }

            // Do second rotation for root
            return (root.left == null) ? root : rightRotate(root);

        } else { // Key lies in the right subtree
            // Key is not in tree, we are done
            if (root.right == null) {
                return root;
            }

            // Zig-Zag (Right-Left)
            if (key < root.right.key) {
                // Recursively bring the key as root of right-left
                root.right.left = splay(root.right.left, key);
                // Do first rotation for root.right
                if (root.right.left != null) {
                    root.right = rightRotate(root.right);
                }
            }
            // Zig-Zig (Right-Right)
            else if (key > root.right.key) {
                // Recursively bring the key as root of right-right
                root.right.right = splay(root.right.right, key);
                // Do first rotation for root
                root = leftRotate(root);
            }

            // Do second rotation for root
            return (root.right == null) ? root : leftRotate(root);
        }
    }

    // --- Public API ---

    /**
     * Searches for a key in the tree.
     * This operation performs splaying on the accessed node (or the last accessed
     * node in the search path), moving it to the root.
     *
     * @param key The integer key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key but not modified,
     * maintaining the set property.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. This brings the closest node to the root.
        root = splay(root, key);

        // If key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // If the new key is smaller than the root's key, make the new node
        // the new root. The old root becomes the right child of the new node,
        // and the old root's left subtree becomes the new node's left subtree.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
            root = newNode;
        }
        // If the new key is greater than the root's key, do the symmetric operation.
        else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
            root = newNode;
        }
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        // If tree is empty, do nothing
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (or the last accessed node) to the root
        root = splay(root, key);

        // If key is not present in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right child becomes the new root.
            root = rightSubtree;
        } else {
            // If there is a left child, we need to join the two subtrees.
            // Splay the left subtree to bring its maximum element to its root.
            // Since `key` is guaranteed to be greater than any element in the
            // left subtree, splaying for `key` will bring the max element to the top.
            leftSubtree = splay(leftSubtree, key);

            // The new root of the left subtree (the max element) will now have
            // a null right child. We attach the original right subtree there.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}